contracts/Witch.sol

```diff
@@ -13,198 +13,358 @@ import "@yield-protocol/utils-v2/contracts/math/WDivUp.sol";
+        Line storage line = lines[vault.ilkId][series.baseId];
+        uint128 art = uint256(balances.art).wmul(line.proportion).u128();
+        if (art < limits_.dust) art = balances.art;
+        uint128 ink = (art == balances.art) ? balances.ink : uint256(balances.ink).wmul(line.proportion).u128();
 
-        limits[vault_.ilkId] = limits_;
         auctions[vaultId] = Auction({
-            owner: vault_.owner,
-            start: block.timestamp.u32()
+            owner: vault.owner,
+            start: block.timestamp.u32(),
+            baseId: series.baseId,
+            art: art,
+            ink: ink
         });
+
+        // The Witch is now in control of the vault under auction
+        // TODO: Consider using `stir` to take only the part of the vault being auctioned.
         cauldron.give(vaultId, address(this));
         emit Auctioned(vaultId, block.timestamp.u32());
     }
 
-    /// @dev Pay `base` of the debt in a vault in liquidation, getting at least `min` collateral.
-    /// Use `payAll` to pay all the debt, using `buy` for amounts close to the whole vault might revert.
-    function buy(bytes12 vaultId, uint128 base, uint128 min)
-        external
-        returns (uint256 ink)
-    {
-        require (auctions[vaultId].start > 0, "Vault not under auction");
-        DataTypes.Vault memory vault_ = cauldron.vaults(vaultId);
-        DataTypes.Series memory series_ = cauldron.series(vault_.seriesId);
-        DataTypes.Balances memory balances_ = cauldron.balances(vaultId);
-        require (balances_.art > 0, "Nothing to buy");                                      // Cheapest way of failing gracefully if given a non existing vault
-
-        Auction memory auction_ = auctions[vaultId];
-        Ilk memory ilk_ = ilks[vault_.ilkId];
-        Limits memory limits_ = limits[vault_.ilkId];
-
-        uint256 art = cauldron.debtFromBase(vault_.seriesId, base);
-        {
-            uint256 elapsed = uint32(block.timestamp) - auction_.start;                      // Auctions will malfunction on the 7th of February 2106, at 06:28:16 GMT, we should replace this contract before then.
-            uint256 price = inkPrice(balances_, ilk_.initialOffer, ilk_.duration, elapsed);
-            ink = uint256(art).wmul(price);                                                    // Calculate collateral to sell. Using divdrup stops rounding from leaving 1 stray wei in vaults.
-            require (ink >= min, "Not enough bought");
-            require (art == balances_.art || balances_.ink - ink >= limits_.dust * (10 ** limits_.dec), "Leaves dust");
-            limits[vault_.ilkId].sum -= ink.u128();
-        }
+    /// @dev Pay `base` of the debt in a vault in liquidation, getting at least `minInkOut` collateral.
+    /// Use `payAll` to pay all the debt, using `paySome` for amounts close to the whole vault might revert.
```

Comment:
`payAll` doesn't exist anymore, I think the msg should be similar to the `payWithTokens` one, so the liquidator just gives enough allowance to buy all knowing he won't overpay